pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AsofJoinOperator.java

```diff
@@ -0,0 +1,183 @@
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.query.planner.logical.RexExpression;
+import org.apache.pinot.query.planner.partitioning.KeySelector;
+import org.apache.pinot.query.planner.partitioning.KeySelectorFactory;
+import org.apache.pinot.query.planner.plannode.JoinNode;
+import org.apache.pinot.query.runtime.blocks.MseBlock;
+import org.apache.pinot.query.runtime.plan.OpChainExecutionContext;
+
+
+public class AsofJoinOperator extends BaseJoinOperator {
+  private static final String EXPLAIN_NAME = "ASOF_JOIN";
+
+  private final Map<Object, NavigableMap<Comparable<?>, Object[]>> _rightTable;
+  private final KeySelector<?> _leftKeySelector;
+  private final KeySelector<?> _rightKeySelector;
+  private final MatchConditionType _matchConditionType;
+  private final int _leftMatchKeyIndex;
+  private final int _rightMatchKeyIndex;
+
+  public AsofJoinOperator(OpChainExecutionContext context, MultiStageOperator leftInput, DataSchema leftSchema,
+      MultiStageOperator rightInput, JoinNode node) {
+    super(context, leftInput, leftSchema, rightInput, node);
+    Preconditions.checkState(node.getNonEquiConditions().isEmpty(),
+        "ASOF_JOIN operator cannot have non-equi join keys");
+    Preconditions.checkState(node.getMatchCondition() != null, "ASOF_JOIN operator must have a match condition");
+    _rightTable = new HashMap<>();
+    List<Integer> leftKeys = node.getLeftKeys();
+
+    _leftKeySelector = KeySelectorFactory.getKeySelector(leftKeys);
+    _rightKeySelector = KeySelectorFactory.getKeySelector(node.getRightKeys());
+
+    RexExpression matchCondition = node.getMatchCondition();
+    Preconditions.checkState(matchCondition instanceof RexExpression.FunctionCall,
+        "ASOF_JOIN operator only supports function call match condition");
+
+    try {
+      _matchConditionType =
+          MatchConditionType.valueOf(((RexExpression.FunctionCall) matchCondition).getFunctionName().toUpperCase());
+    } catch (IllegalArgumentException e) {
+      throw new UnsupportedOperationException("Unsupported match condition: " + matchCondition);
+    }
+    List<RexExpression> matchKeys = ((RexExpression.FunctionCall) matchCondition).getFunctionOperands();
+    // TODO: Add support for MATCH_CONDITION containing two columns of different types. In that case, there would be
+    //       a CAST RexExpression.FunctionCall on top of the RexExpression.InputRef, and we'd need to add the
+    //       appropriate type casts to make sure that the Comparable based comparisons in this class don't throw
+    //       ClassCastException.
+    Preconditions.checkState(
+        matchKeys.size() == 2 && matchKeys.get(0) instanceof RexExpression.InputRef
+            && matchKeys.get(1) instanceof RexExpression.InputRef,
+        "ASOF_JOIN operator only supports match conditions with a comparison between two columns of the same type");
```

Comment:
How is this checking that both columns are of the same type? I'm only seeing it checking they are instances of type `InputRef`,  but maybe I'm missing something.

Reply:
The TODO above mentions the reason - basically, if the two columns are of different but coercible types, Calcite will introduce a cast function call on one side of the match condition. So one of the keys will no longer be an `InputRef`, but a cast `FunctionCall` (with the `InputRef` being its first operand).